### PR TITLE
feat(query): binary join for metrics across multiple partitions and having shard key regex

### PR DESCRIPTION
**Pull request by TanviBhavsar**
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?

Support for binary join for metrics across multiple partitions and having queries with shard key regex

### CODE REVIEW

File: coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala [hunk 1]
```diff
@@ -131,16 +161,44 @@ class MultiPartitionPlanner(partitionLocationProvider: PartitionLocationProvider
...
+    }
+
+    val onKeysReal = ExtraOnByKeysUtil.getRealOnLabels(logicalPlan, queryConfig.addExtraOnByKeysTimeRanges)
+
+    if (logicalPlan.operator.isInstanceOf[SetOperator])
+      SetOperatorExec(qContext, InProcessPlanDispatcher, Seq(lhsExec), Seq(rhsExec), logicalPlan.operator,
+        LogicalPlanUtils.renameLabels(onKeysReal, datasetMetricColumn),
+        LogicalPlanUtils.renameLabels(logicalPlan.ignoring, datasetMetricColumn), datasetMetricColumn)
+    else
+      BinaryJoinExec(qContext, InProcessPlanDispatcher, Seq(lhsExec), Seq(rhsExec), logicalPlan.operator,
```
Comment (vishramachandran) [inline coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala:187]: Let's discuss offline if InProcessPlanDispatcher is the right one for this.

File: coordinator/src/main/scala/filodb.coordinator/queryplanner/ShardKeyRegexPlanner.scala [hunk 2]
```diff
@@ -81,11 +84,29 @@ class ShardKeyRegexPlanner(dataset: Dataset,
...
+    val rhsExec = materialize(logicalPlan.rhs, rhsQueryContext)
+
+    val onKeysReal = ExtraOnByKeysUtil.getRealOnLabels(logicalPlan, queryConfig.addExtraOnByKeysTimeRanges)
+
+    val execPlan = if (logicalPlan.operator.isInstanceOf[SetOperator])
+         SetOperatorExec(qContext, InProcessPlanDispatcher, Seq(lhsExec), Seq(rhsExec), logicalPlan.operator,
+          LogicalPlanUtils.renameLabels(onKeysReal, datasetMetricColumn),
+          LogicalPlanUtils.renameLabels(logicalPlan.ignoring, datasetMetricColumn), datasetMetricColumn)
+      else
+         BinaryJoinExec(qContext, InProcessPlanDispatcher, Seq(lhsExec), Seq(rhsExec), logicalPlan.operator,
```
Comment (vishramachandran) [inline coordinator/src/main/scala/filodb.coordinator/queryplanner/ShardKeyRegexPlanner.scala:105]: Need to think more about InProcessPlanDispatcher. lets discuss.

File: coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala [hunk 3]
```diff
@@ -78,26 +81,53 @@ class MultiPartitionPlanner(partitionLocationProvider: PartitionLocationProvider
...
-      sortBy(_.timeRange.startMs)
-    if (partitions.isEmpty) new UnsupportedOperationException("No partitions found for routing keys: " + routingKeyMap)
+    val partitions = if (routingKeys.isEmpty) List.empty
+    else {
+      val routingKeyMap = routingKeys.map(x => (x._1, x._2.head)).toMap
+      partitionLocationProvider.getPartitions(routingKeyMap, queryTimeRange).
+        sortBy(_.timeRange.startMs)
+    }
+    if (partitions.isEmpty && !routingKeys.isEmpty)
+      new UnsupportedOperationException("No partitions found for routing keys: " + routingKeys)
```
[the remark this answers is not in the document]
Comment (TanviBhavsar) [reply coordinator/src/main/scala/filodb.coordinator/queryplanner/MultiPartitionPlanner.scala:91]: If routingKeys.isEmpty partitions would be empty. We don't need to throw an exception as it is expected. Materialize function will use localplanner to materialize the logical plan when routing keys are empty and we don't need the partitions

**Review by tjackpaul — APPROVED** [review]
lgtm.. thank you @TanviBhavsar

**Pull request merged**